crypto/http/http_lib.c

```diff
@@ -59,44 +60,70 @@ int OSSL_HTTP_parse_url(const char *url, char **phost, char **pport,
             if (pssl != NULL)
                 *pssl = 1;
             port = OSSL_HTTPS_PORT;
+            portnum = 443;
         } else if (strcmp(buf, OSSL_HTTP_NAME) != 0) {
             goto parse_err;
         }
 
-        /* Check for double slash */
+        /* check for double slash */
         if ((p[0] != '/') || (p[1] != '/'))
             goto parse_err;
         p += 2;
     }
     host = p;
 
-    /* Check for trailing part of path */
-    p = strchr(p, '/');
-    if (ppath != NULL && (*ppath = OPENSSL_strdup(p == NULL ? "/" : p)) == NULL)
-        goto err;
-    if (p != NULL)
-        *p = '\0'; /* Set start of path to 0 so hostname[:port] is valid */
-
-    p = host;
+    /* parse host name/address as far as needed here */
     if (host[0] == '[') {
-        /* ipv6 literal */
+        /* ipv6 literal, which may include ':' */
         host++;
-        p = strchr(host, ']');
-        if (p == NULL)
+        host_end = strchr(host, ']');
+        if (host_end == NULL)
             goto parse_err;
-        *p = '\0';
-        p++;
+        *host_end++ = '\0';
+    } else {
+        host_end = strchr(host, ':');
+        if (host_end == NULL)
+            host_end = strchr(host, '/');
+        if (host_end == NULL)
+            host_end = host + strlen(host);
     }
 
-    /* Look for optional ':' for port number */
-    if ((p = strchr(p, ':'))) {
-        *p = '\0';
-        port = p + 1;
+    /* parse optional port specification starting with ':' */
+    p = host_end;
+    if (*p == ':') {
+        port = ++p;
+        if (pport_num == NULL) {
+            p = strchr(port, '/');
+            if (p == NULL)
+                p = p + strlen(port);
+        } else { /* make sure a numerical port value is given */
+            portnum = strtol(port, &p, 10);
+            if (p == port || (*p != '\0' && *p != '/'))
+                goto parse_err;
+            if (portnum < 1 || portnum > 65535) {
+                HTTPerr(0, HTTP_R_INVALID_PORT_NUMBER);
+                goto err;
+            }
+        }
     }
+    *host_end = '\0';
+    *p = '\0'; /* terminate port string */
+
+    /* check for optional path at end of url starting with '/' */
+    path = url + (p - buf);
```

Comment:
Not sure why you do this..
Could you just have done this?

```
if (*p == '\') {
    *p = '\0';
    path = p + 1;
}
```

Reply:
I've added a comment explaining this:
```
    /* cannot use p + 1 because *p is '\0' and path must start with '/' */
```

Reply:
??? If you dont set it to '\0' above the if then you dont have this problem..
i.e without you setting it it will be either '\' or '\0'
if it is '\' then you can then do as I have above otherwise it is at the end.

Reply:
Yes, but what you are missing is that even if `*p = '\0'` there is still the problem that
the path output needs needs to be guaranteed to start with the preceding `/`
but in your code it starts with the character thereafter, and
we cannot both replace the `/` by `\` and keep the  `/` in the same piece of memory.